blog/2022/storage-map.adoc

```diff
@@ -0,0 +1,127 @@
+
+### Next step: Persist data in relational database
+
+_Requirements:_ Postgres 14 or CockroachDB 22.1 (or newer)
+
+To keep the data in a database, run the following command:
+
+   bin/kc.sh start-dev --storage=jpa --db-url=<jdbc-url> --db-username=<username> --db-password=<password>
+
+Reference to https://github.com/keycloak/keycloak/discussions/9596[Q&A live document].
+
+CAUTION: This feature has known issues in CockroachDB, see https://github.com/keycloak/keycloak/issues/11618[this tracker for list].
+
+### Alternative next step: Persist data in external Infinispan
+
+_Requirements:_ Infinispan 12 (or newer)
+
+If you have a cluster of running Infinispan, you can use it for storing the data as well,
+just run the following command:
+
+   bin/kc.sh start-dev --storage=hotrod --storage-hotrod-host=<host> --storage-hotrod-port=11222 --storage-hotrod-username=<username> --storage-hotrod-password=<password>
+
+Keycloak creates the needed caches in Infinispan upon first start if the caches
+do not exist already. You are free to create and configure the caches
+yourself if you prefer to fine-tune those or e.g. want to set up persistence.
+Refer to Infinispan documentation for available options for cache configuration.
+
+Reference to https://github.com/keycloak/keycloak/discussions/12010[Q&A live document].
+
+=== Next step: Store data in separate storages
+
+Eventually it turns out that some of the data should be stored in files or a database, other
+(e.g. session) data should be stored in Infinispan. Can this be achieved?
+
+Yes! Keycloak storage is divided into the following storage areas: Realms,
+clients, users, groups, roles, client scopes, authorization services,
+events, authentication sessions, user/client sessions, login failures,
+action tokens, and single-use tokens (last two areas to be merged in Keycloak 20).
+
+NOTE: For more details on this division, please see https://github.com/keycloak/keycloak-community/blob/main/design/keycloak.x/storage-architecture.md[architecture specification].
+
+Each area maintains its own storage for storing the data, and each area
+is independent of others. In other words, realms can be served by database and
+users by Infinispan.
+
+To have all the session data stored in external Infinispan,
+and realm / client / user / group / role / ... data stored in a relational database, you
+can issue the following command:
+
+   bin/kc.sh start-dev --storage=jpa \
+     --db-url=<jdbc-url> --db-username=<username> --db-password=<password> \
+     --storage-hotrod-host=<host> --storage-hotrod-port=11222 \
+     --storage-hotrod-username=<username> --storage-hotrod-password=<password> \
+     --storage-area-action-token=hotrod \
+     --storage-area-auth-session=hotrod \
+     --storage-area-single-use-object=hotrod \
+     --storage-area-user-session=hotrod
+
+This starts Keycloak server wit all areas being handles by relational database
```

Comment:
Typo, "handles" should be "handled" IMHO

Reply:
True!